### ISSUE TITLE
[Feature] Formatting & Item Exporting

### ISSUE THREAD

**Issue by Flavourized** [JustBru00/RenamePlugin#86]
Hallo there

I wanted to request a feature which will take an extra step on top of the normal formatting this plugin has to offer. 

-----------------------------------------
**Pushed to separate issue #103. Read why from within that issue.** 
1. Adding ~~a centered command~~ an alignment command
- If i have an item all setup with my lores and item names and all, and want to center them in the tooltip, this command will do that. 
- The alignment should be based off the longest line in the tooltip, so that the centering will look lined up

Command: 
`<required arg>` `[optional arg]`
`/align name <left || center || right>`
`/align lore <left || center || right> <line0,line1,lineN...>`

- Example: 
  + Initial item here: http://prntscr.com/ijbcuq
  + Type `/align name center`
  + Type `/align lore center 1,2`
  + Final item: http://prntscr.com/ijbex9

-----------------------------------------
**Pushed to issue #69**
2. Adding commands for manipulating ItemMeta
  a. Hide attributes -  Hide item attributes
  b. Hide enchantments (maybe hook with TokenEnchant/CrazyEnchant/etc to hide custom enchants)
  c. Hide Effects - Hide item effect attribute
  d. Hide Unbreakable - Hide item unbreakable attributes

Commands (Toggle commands):
`<required arg>`  `[optional arg]`
`/hide attributes`
`/hide enchantments`
`/hide effects`
`/hide unbreakable`

-----------------------------------------
**Pushed into issue #104**
3. Add (mvdw)PlaceholderAPI support, so we can put placeholders in our items

Command:
- No special command, just include the placeholder in `/rename` or `/lore`, etc.. 

-----------------------------------------
**Pushed into issue #105 and #106**
4. Add command to export item
- This command will export the item or inventory of the player to a file in /plugins/epicrename/items .
- Possible alternative to upload to pastebin/hastebin/cl1p/etc

Commands: 
`<required arg>`  `[optional arg]`
`/export <hand || inventory>`

Export format: 
```
Items:
  <inventorySlotNumber>:
    material: MATERIAL_NAME
    data: 0
    amount: 1
    display_name: "&f%player_name%'s example item"
    lore:
    - "&f  Lore line 0!"
    - "&7This is lore line 1"
    enchantments:
    - 'enchant1:level'
    - 'enchant2:level'
    hide_attributes: <true/false>
    hide_enchantments: <true/false>
    hide_effects: <true/false>
    hide_unbreakable: <true/false>
```

Export format Example:
Item:  http://prntscr.com/ijbqji (pretend this sword is in slot 35 (bottom right slot of inv))
1. Type `/export hand`

```
Items:
  '35':
    material: DIAMOND_SWORD
    data: 0
    amount: 1
    display_name: '           &a&lSWORDS'
    lore:
    - '&f'
    - '   &7&oClick to see all custom &f '
    - '&f &7&oenchantments for &a&oswords&7&o! '
    enchantments:
    - 'DURABILITY:1'
    - 'SHARPNESS:1"
    hide_attributes: true
    hide_enchantments: true
    hide_effects: true
    hide_unbreakable: true
```

-----------------------------------------

I think these features would be pretty useful, and makes it a lot easier for server owners to add some form of quality to their servers

**Comment by JustBru00** [comment]
Thanks for such a detailed request. It really helps for me to know what you are looking for.

- The centered command sounds great. I don't know why I have never thought of it before.
- Adding commands for ItemMeta data has been partially requested before in #69. It should be a great addition. I am undecided on whether it will be /hide or separate commands such as /hideunbreakable.
- PlaceholderAPI support might be possible. Allowing the placeholder in the command shouldn't be too hard. Unfortunately, updating the placeholders will be difficult. 
- Exporting the item and allowing it to be imported again has been suggested in #29. Your idea sounds much better instead of /saveitem and /getitem. I like your idea of the config style for the items, although, I have an idea to use some sort of method that shortens the item to like a 64 character string. If that is not possible I may try to create a server that allows you to preform the command /export <item,inventory>. After you preform the command it sends the item data to a server and returns a random URL. Something like http://rename.epicrealmmc.com/Fzt0nuXonPc. Then you can use /import <URL>. I might be able to use pastebin for that feature instead of a custom server.

**Comment by JustBru00** [comment]
Currently researching possibility of adding PlaceholderAPI support.

**Comment by Flavourized** [comment]
> PlaceholderAPI support might be possible. Allowing the placeholder in the command shouldn't be too hard. Unfortunately, updating the placeholders will be difficult.

I think updating placeholder in name/lore will be a bit difficult/unnecessary, but that is just my opinion. 

The way i am thinking of, is that you would need to iterate over each item in the player's inventory whenever some event is triggered (join event seems the most plausible). If a placeholder needs updating then do it that way, but i think that will cause a lot of lag if it is going to be done automatically. And surely wont work for servers with 50+ players even.

The other way i am thinking, is to make it manual, where a command is typed. And only when that command is typed, it will updated whichever placeholders are in that player's inventory. 

I think the ability to put placeholders in the name/lore which dont update would be enough. Possibly just for things where you can put the player's name in the name/lore.

**Comment by JustBru00** [comment]
Would you want the placeholders to update per player. Thinking about the linguistics of it, I don;t know if it can be updated at all after the first command is run. My plugin will have no idea what is a placeholder anymore.

**Comment by Flavourized** [comment]
>  My plugin will have no idea what is a placeholder anymore.

Yes i was just thinking of this. After the placeholder is used in the item on /rename or /lore, it will permanently change on the item.

The only way around this would be to create an entire item management system, which i dont think would fit very nicely with this plugin

**Comment by JustBru00** [comment]
> The only way around this would be to create an entire item management system, which i dont think would fit very nicely with this plugin

I agree, that would really get outside of this plugin's scope.

**Comment by JustBru00** [comment]
So, you just want them to work the first time. That I can do.

**Comment by Flavourized** [comment]
p.s i know there are some inconsistencies in my post, but i have a feeling that you understand what i meant. I can fix if you need me to

**Comment by JustBru00** [comment]
> p.s i know there are some inconsistencies in my post, but i have a feeling that you understand what i meant. I can fix if you need me to

I think I got it. If I have any further questions I will ask. Thanks again for being so detailed.

**Comment by Flavourized** [comment]
https://gist.github.com

**Comment by JustBru00** [comment]
> https://gist.github.com

Uhhh... Did you post the wrong link?

**Comment by Flavourized** [comment]
Not quite, but I forgot to type my message cause I'm a noob.

I was gonna say you could use gist as well. This is what essX uses with their in-game kit creator. 

We type /createkit, and then the plugin will produce a gist url.

**Comment by JustBru00** [comment]
Yeah, I thought about that. I will have to research the post limit for both pastebin and gist.

**Comment by JustBru00** [comment]
Note to self: 

Possibly /hideitem <stuff>
Example: /hideitem unbreakable

**Comment by Flavourized** [comment]
Any chance these features are coming?

**Comment by JustBru00** [comment]
I am working on them when I can. I have started working full time so I don't have near as much free time to work on it.
I wanted to get an update for 1.13 out as soon as I could, so I had to wait on releasing these features.

**Comment by JustBru00** [comment]
```
align:
  name_aligned_left_success: '&aName aligned to the left.'
  name_aligned_center_success: '&aName aligned to the center.'
  name_aligned_right_success: '&aName aligned to the right.'
  lore_aligned_left_success: '&aLore aligned to the left.'
  lore_aligned_center_success: '&aLore aligned to the center.'
  lore_aligned_right_success: '&aLore aligned to the right.'
  both_aligned_left_success: '&aName and lore aligned to the left.'
  both_aligned_center_success: '&aName and lore aligned to the center.'
  both_aligned_right_success: '&aName and lore aligned to the right.'
  incorrect_name_args: '&cIncorrect arguments /align name <left, center, right>'
  incorrect_lore_args: '&cIncorrect arguments /align lore <left, center, right>'
  incorrect_both_args: '&cIncorrect arguments /align both <left, center, right>'
  incorrect_args: '&cIncorrect arguments /align <name, lore, both> <left, center, right>'
  not_enough_args: '&cIncorrect arguments /align <name, lore, both> <left, center, right>'
  disabled_world: '&cYou are in a disabled world.'
  wrong_sender: '&cSorry only a player can use that command.'
  no_permission: '&cSorry you don''t have permission to perform that command.'

```

/align messages.yml text.
Removed for temporary as of version 3.5-INDEV

**Comment by JustBru00** [comment]
/align pushed into separate issue due to problems with font character size.

**Comment by JustBru00** [comment]
I am going to close this issue as the features that have not been completed yet have their own issues: #103 #69 #104
